crates/net/eth-wire/src/disconnect.rs

```diff
@@ -143,6 +148,45 @@ impl Decodable for DisconnectReason {
     }
 }
 
+/// This trait is meant to allow higher level protocols like `eth` to disconnect from a peer, using
+/// lower-level disconnect functions (such as those that exist in the `p2p` protocol) if the
+/// underlying stream supports it.
+#[async_trait::async_trait]
+pub trait CanDisconnect<T>: Sink<T> + Unpin + Sized {
+    /// Disconnects from the underlying stream, using a [`DisconnectReason`] as disconnect
+    /// information if the stream implements a protocol that can carry the additional disconnect
+    /// metadata.
+    async fn disconnect(
+        &mut self,
```

Comment:
using mut borrow is fine I guess, mimicks what `TcpStream::shutdown` does as well.

and we likely need this mut borrow if we want to use it in Sink/Stream impls